### PR TITLE
fix(pnp): preserve trailing slashes inside the packageLocations of the serialized state

### LINKED ISSUE #1521: [Bug] can't require local workspace in folder named "package" next to a package.json
- [x] I'd be willing to implement a fix

**Describe the bug**

I was trying to create a repro for an issue with workspace dependencies, so I set up a single child workspace in a folder called `package`. It is impossible to require this package via its name if the package uses the implicit `index.js` main entry.

**To Reproduce**

In an empty project, create a workspace named `my-pkg` in a folder named `package`. Add an `index.js` file, but don't add it to `my-pkg`'s package.json as main.
Resolving the package returns the path to the project's root package.json (i.e. the package.json next to the `my-pkg` package's folder called `package`).

<details>
  <summary>Reproduction</summary>

  ```js repro
  const {dirname, resolve} = require('path');
  const {promises: {writeFile, mkdir}} = require('fs');

  await mkdir('package');
  await writeFile('package/package.json', JSON.stringify({
    name: 'my-pkg',
    version: '0.1.0',
    // main: 'index.js', <-- no main!
  }, null, 2));

  await writeFile('package/index.js', `
  // this file has to exist
  `);

  await packageJsonAndInstall({
    private: true,
    workspaces: ['package'],
    dependencies: {
      [`my-pkg`]: `workspace:package`,
    },
  });

  const packageJsonPath = await node(`require.resolve('my-pkg/package.json')`);

  await expect(node(`require.resolve('my-pkg')`)).resolves.toBe(
    resolve(dirname(packageJsonPath), 'index.js')
  );
  ```
</details>

### PR DESCRIPTION
**Pull request by paul-soporan**
**What's the problem this PR addresses?**

Trailing slashes part of the `packageLocations` of the PnP serialized state weren't preserved by `hydrateRuntimeState` and `resolveToUnqualified` when resolving packages (because `ppath.resolve` was used, which doesn't preserve trailing slashes), which caused the PnP unqualified resolution to return files instead of folders.

Fixes #1521.

**How did you fix it?**

I replaced `ppath.resolve` with `ppath.join` (I made sure it was safe - the first path always being absolute and the second one relative), which does preserve trailing slashes.

I also added a regression test.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I have verified that all automated PR checks pass.

### CODE REVIEW

**Comment by paul-soporan** [comment]
Side note: I also made it so that the candidates are normalized inside the unqualified resolution error, so that they don't looks as weird as before with trailing slashes:

image

File: packages/yarnpkg-pnp/sources/loader/makeApi.ts [hunk 1]
```diff
@@ -734,7 +738,7 @@ export function makeApi(runtimeState: RuntimeState, opts: MakeApiOptions): PnpAp
     } else {
       throw makeError(
         ErrorCode.QUALIFIED_PATH_RESOLUTION_FAILED,
-        `Qualified path resolution failed - none of the candidates can be found on the disk.\n\nSource path: ${unqualifiedPath}\n${candidates.map(candidate => `Rejected candidate: ${candidate}\n`).join(``)}`,
+        `Qualified path resolution failed - none of the candidates can be found on the disk.\n\nSource path: ${unqualifiedPath}\n${candidates.map(candidate => `Rejected candidate: ${ppath.normalize(candidate)}\n`).join(``)}`,
```
Comment (merceyz) [inline packages/yarnpkg-pnp/sources/loader/makeApi.ts:741]: Shouldn't this be converted to a native path?

Comment (paul-soporan) [reply packages/yarnpkg-pnp/sources/loader/makeApi.ts:741]: Good catch! I'll update all of the paths inside errors.

File: packages/plugin-node-modules/sources/NodeModulesLinker.ts [hunk 2]
```diff
@@ -395,7 +395,10 @@ const parseLocation = (location: PortablePath, {skipPrefix}: {skipPrefix: Portab
   if (projectRelativePath === null)
     throw new Error(`Assertion failed: Cannot process a path that isn't part of the requested prefix (${location} isn't within ${skipPrefix})`);
 
-  const allSegments = projectRelativePath.split(ppath.sep);
+  const allSegments = projectRelativePath
+    .split(ppath.sep)
+    // Ignore empty segments (after trailing slashes)
+    .filter(Boolean);
```
Comment (arcanis) [inline packages/plugin-node-modules/sources/NodeModulesLinker.ts:401]:
```suggestion
    .filter(segment => segment !== ``);
```

**Pull request merged**